go/host/enclave/service.go

```diff
@@ -200,27 +175,108 @@ func (e *Service) Unsubscribe(id rpc.ID) error {
+// exactly one active sequencer enclave for that.
+// It will promote a new active sequencer if the current one fails or is unavailable.
+func (e *Service) managePeriodicBatches() {
+	interval := e.batchInterval
+	if interval == 0 {
+		interval = 1 * time.Second
+	}
+	// we use ticks rathen than sleeps to maintain batch production cadence
+	batchProductionTicker := time.NewTicker(interval)
+
+	for e.running.Load() {
+		// wait for the next tick
+		<-batchProductionTicker.C
+
+		activeSeq, err := e.getActiveSequencerGuardian()
+		if err != nil {
+			e.logger.Info("No active sequencer found, trying to promote a new one", log.ErrKey, err)
+			e.tryPromoteNewSequencer()
+			continue
+		}
+
+		if activeSeq.InSyncWithL1() {
+			err = activeSeq.ProduceBatch()
 			if err != nil {
-				e.logger.Info("Unable to promote new sequencer.", log.EnclaveIDKey, enclID, log.ErrKey, err)
+				// todo: do we want to have a few failed attempts before looking to promote a new one?
+				e.logger.Error("Sequencer failed to produce batch", log.ErrKey, err)
+				// if the active sequencer fails, we will try to promote a new one
+				e.tryPromoteNewSequencer()
 				continue
 			}
-			e.activeSequencerID.Store(enclID)
-			e.logger.Warn("Successfully promoted new sequencer.", log.EnclaveIDKey, enclID)
-			return
 		}
-		// wait for retry interval before trying again, enclaves may not be ready yet or may be awaiting permissioning
-		time.Sleep(_promoteSeqRetryInterval)
+
+		// sanity check: make sure there are no enclaves, which think they are active sequencers
+		for _, guardian := range e.enclaveGuardians {
+			if guardian.state.IsEnclaveActiveSequencer() && guardian != activeSeq {
+				e.logger.Warn("Found unexpected active sequencer in guardians, demoting...",
+					log.EnclaveIDKey, guardian.GetEnclaveID(), "expectedActiveID", activeSeq.GetEnclaveID().Hex())
+				guardian.DemoteFromActiveSequencer()
+			}
+		}
 	}
+	batchProductionTicker.Stop()
+	e.logger.Info("Stopping periodic batch production.")
 }
 
+func (e *Service) tryPromoteNewSequencer() {
+	prevActiveSeqID := e.activeSequencerID.Load()
+
+	// prepare a list of guardians to try, starting with the one that thinks it is active (if there is one, e.g. after host restart)
+	// and skipping the one that just got demoted (if there was one)
+	guardiansToTry := make([]*Guardian, 0, len(e.enclaveGuardians))
+
+	for _, guardian := range e.enclaveGuardians {
+		enclID := guardian.GetEnclaveID()
+		if prevActiveSeqID != nil && *enclID == *prevActiveSeqID {
+			// skip this enclave since we are replacing it
+			continue
+		}
+
+		// if the network is just starting up the candidate just needs to be in sync with L1
+		networkStartingUp := e.sl.L2Repo().FetchLatestBatchSeqNo().Cmp(big.NewInt(1)) < 0
+		if networkStartingUp && guardian.InSyncWithL1() || guardian.IsLive() {
+			if guardian.state.IsEnclaveActiveSequencer() {
```

Comment:
I'm a bit confused by this logic. First we skip the guardian that was active, but we also prepend it when it's active. Is this a way to handle when the state of the "service" falls out of sync with the "guardian"?

Reply:
a comment about the startup logic is useful. The state of the components at startups and what we want to do if all goes well